src/daemon/daemon.cpp

```diff
@@ -386,10 +395,91 @@ auto connect_rpc(mp::DaemonRpc& rpc, mp::Daemon& daemon)
+    QObject::connect(&rpc, &mp::DaemonRpc::on_restart, &daemon, &mp::Daemon::restart, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_delete, &daemon, &mp::Daemon::delet, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_umount, &daemon, &mp::Daemon::umount, Qt::BlockingQueuedConnection);
     QObject::connect(&rpc, &mp::DaemonRpc::on_version, &daemon, &mp::Daemon::version, Qt::BlockingQueuedConnection);
 }
+
+template<typename Targets, typename InstanceMap>
+grpc::Status validate_targets(const Targets& targets,
+                              const InstanceMap& vms,
+                              const InstanceMap& deleted)
+{
+    fmt::memory_buffer errors;
+    for(const auto& name : targets)
+        if(vms.find(name) == std::cend(vms))
+        {
+            if(deleted.find(name) == std::cend(deleted))
+                fmt::format_to(errors, "instance \"{}\" does not exist\n", name);
+            else
+                fmt::format_to(errors, "instance \"{}\" is deleted\n", name);
+        }
+
+    return errors.size() ? grpc_status_for(errors) : grpc::Status::OK;
+}
+
+template<typename Targets, typename InstanceMap>
+auto process_targets(const Targets& targets,
+                     const InstanceMap& vms,
+                     const InstanceMap& deleted)
+-> std::pair<std::vector<typename Targets::value_type>, grpc::Status>
+{   // TODO: use this in commands that currently duplicate the same kind of code
+    auto st = validate_targets(targets, vms, deleted);
+    auto tgts = std::vector<typename Targets::value_type>{};
+
+    if(st.ok())
+    {
+        if(targets.empty())
+            for(const auto& vm_item : vms)
+                tgts.push_back(vm_item.first);
+        else
+            std::copy(std::cbegin(targets), std::cend(targets),
+                      std::back_inserter(tgts));
+    }
+
+    return std::make_pair(tgts, st);
+}
+
+mp::SSHProcess exec_and_log(mp::SSHSession& session, const std::string& cmd)
+{
+    mpl::log(mpl::Level::debug, category, fmt::format("Executing {}.", cmd));
+    return session.exec(cmd);
+}
+
+grpc::Status stop_accepting_ssh_connections(mp::SSHSession& session)
+{
+    auto proc = exec_and_log(session, stop_ssh_cmd);
+    auto ecode = proc.exit_code();
+
+    return ecode == 0 ? grpc::Status::OK
+                      : grpc::Status{grpc::StatusCode::FAILED_PRECONDITION,
+                                     fmt::format("Could not stop sshd. '{}' exited with code {}", stop_ssh_cmd, ecode),
+                                     proc.read_std_error()};
+}
+
+grpc::Status ssh_reboot(const std::string& hostname, int port,
+                        const std::string& username, const mp::SSHKeyProvider& key_provider)
+{
+    mp::SSHSession session{hostname, port, username, key_provider};
+    stop_accepting_ssh_connections(session); /* This allows us to later detect when the machine
```

Comment:
Could you please put the comment above `stop_accepting_ssh_connections(session);` and use C++ style comments?

Reply:
Done.